Apache-Beam-Jobs.md

```diff
@@ -227,6 +229,257 @@ error_pcoll = (
+
+We use `ndb_io.GetModels()` because we want to work on `PCollection`s of models, not a list of models. In fact, all operations that can be taken on models (`get`, `put`, `delete`) have analogous `PTransform` interfaces defined in `ndb_io`. They are:
+
+| NDB function           | `PTransform` analogue                        |
+| ---------------------- | -------------------------------------------- |
+| `models = get_multi()` | `model_pcoll = ndb_io.GetModels(Query(...))` |
+| `put_multi(models)`    | `model_pcoll \| ndb_io.PutMulti()`           |
+| `delete_multi(keys)`   | `key_pcoll \| ndb_io.DeleteMulti()`          |
+
+Note that `get_multi` has the biggest change in interface, in that it takes a `Query` argument. You can get a query for any model by using the class method `get_all`.
+-   **IMPORTANT:** Never use `datastore_services.query_everything()`!! Due to a limitation in Apache Beam, this operation is incredibly slow and inefficient! **You are almost certainly doing something wrong if you need this function.** Ask @brianrodri/@vojtechjelinek for help if you believe you need to use it regardless.
+
+Why should we use these `PTransform` over the simpler `get`/`put`/`delete` functions? **Performance**. The `get`/`put`/`delete` function calls are all *synchronous*, so your job's performance will suffer greatly by waiting for the operations to complete.
+
+`PTransform`s, on the other hand, are specially crafted to take advantage of the Apache Beam framework and guarantee better performance. In general, **you should always prefer `ndb_io` over any of the `get`/`put`/`delete` functions!** If you think you have a valid need for avoiding `ndb_io`, then speak with @brianrodri/@vojtechjelinek first.
+
+Let's get back to implementing the job.
+
+```python
+def run(self):
+    exp_model_pcoll = (
+        self.pipeline
+        | 'Get all ExplorationModels' >> ndb_io.GetModels(
+            exp_models.ExplorationModel.get_all(deleted=False))
+    )
+
+    state_count_pcoll = (
+        exp_model_pcoll
+        | 'Count states' >> beam.Map(self.get_number_of_states)
+    )
+```
+
+Note that we chose to use `beam.Map` here instead of `beam.ParDo`. This is mostly a stylistic choice, as `beam.Map` is just a specialized version of `ParDo`, in that `Map` simply takes each input element and "maps" them to a single output element. In our case, each `ExplorationModel` will map to a single `int`, the number of states.
+
+```python
+def get_number_of_states(self, exp_model: ExplorationModel) -> int:
+    exp = exp_fetchers.get_exploration_from_model(exp_model)
+    return len(exp.states)
+```
+
+Our implementation is simple, we transform the model into a domain object and measure the length of its `states` dictionary.
+
+Finally, we need to sum all the counts together. Fortunately, Apache Beam already has a `PTransform` that does exactly this:
```

Comment:
I don't immediately see what this PTransform in question is when looking at the code. Maybe move the 'CombineGlobally' part of the explanation up here.

Reply:
Done.